torch/distributed/fsdp/fully_sharded_data_parallel.py

```diff
@@ -2539,70 +2569,109 @@ def shard_full_optim_state_dict(
                 ``model.parameters()``. (Default: ``None``)
 
         Returns:
-            Dict[str, Any]: The full optimizer state dict remapped to flattened
-            parameters instead of unflattened parameters and restricted to only
-            include this rank's part of the optimizer state.
+            Dict[str, Any]: The full optimizer state dict now remapped to
+            flattened parameters instead of unflattened parameters and
+            restricted to only include this rank's part of the optimizer state.
         """
-        full_osd = full_optim_state_dict  # alias
-        if "state" not in full_osd or "param_groups" not in full_osd:
-            raise ValueError(
-                "`full_optim_state_dict` must have the keys \"state\" and "
-                "\"param_groups\" to be a valid optimizer state dict"
-            )
+        return _flatten_full_optim_state_dict(
```

Comment:
The flattening logic for `shard_full_optim_state_dict()` and for `scatter_full_optim_state_dict()` on rank 0 are the same, except for the former, only the shard relevant to the rank is kept, while for the latter, each shard is kept, to be broadcast shard-by-shard later.

To this end, I refactored the common logic into `_flatten_full_optim_state_dict()` with a `bool` argument `shard_state`. It is `True` for `shard_full_optim_state_dict()` and `False` for `scatter_full_optim_state_dict()`.